feat(samples): add Unity build pipeline with TeamCity integration

gabebatista opened the pull request:
**Issue number:** N/A

  ## Summary

  ### Changes

  This PR adds a new Unity build pipeline sample demonstrating how to deploy a production-ready Unity CI/CD infrastructure on AWS. The sample integrates Perforce (version control with code review), TeamCity
  (CI/CD), Unity Accelerator (asset caching), and Unity Floating License Server with comprehensive documentation and a Docker image for Unity TeamCity build agents.

  ### User experience

  **Before:** Users had no reference sample showing how to deploy a complete Unity build pipeline on AWS. Individual Unity module documentation existed, but integration patterns were unclear.

  **After:** Users can deploy a production-ready Unity build pipeline with `terraform apply`, following clear step-by-step documentation. The sample demonstrates best practices for integrating all components
  with proper networking, security, and DNS configuration.

  ## Testing This PR

  To test this sample:

  1. **Review documentation**: Read `samples/unity-build-pipeline/README.md` to understand the architecture and deployment steps
  2. **Build Docker image**: Navigate to `samples/unity-build-pipeline/docker/teamcity-unity-build-agent/` and run `./build-and-push.sh` to verify the Unity TeamCity agent image builds successfully (~20
  minutes)
  3. **Validate Terraform**: Run `terraform init` and `terraform validate` in the sample directory
  4. **Review configuration files**: Check `terraform.tfvars.example` and `variables.tf` for proper variable definitions
  5. **Deploy and test infrastructure**: Deploy all components and validate connectivity:
     - Configure `terraform.tfvars` with your Route53 domain, license server zip path, and TeamCity agent image URI
     - Run `terraform apply` to deploy all components
     - Verify all ECS services are healthy in the console
     - Get Perforce super user credentials:
       ```bash
       aws secretsmanager get-secret-value \
         --secret-id $(terraform output -raw perforce_super_user_username_secret_arn) \
         --query SecretString --output text
       aws secretsmanager get-secret-value \
         --secret-id $(terraform output -raw perforce_super_user_password_secret_arn) \
         --query SecretString --output text
       ```
     - Access P4 Swarm (Code Review) at `https://swarm.yourdomain.com` and verify login with Perforce credentials
     - Access TeamCity at the URL from `terraform output -raw teamcity_url`
     - Complete TeamCity initial setup wizard and create an administrator account
     - Verify build agents appear in **Agents → Agents** tab (may take 2-3 minutes):
       - Agents will initially appear in the **Unauthorized** tab
       - Click on the **Unauthorized** tab
       - Click **Authorize** next to each agent
       - Agents will move to the **Connected** tab once authorized
     - Create a test project to validate infrastructure:
       1. Click **Create project**
       2. Name it "Infrastructure Test"
       3. Click **Create** and then **Skip VCS root creation** (no repository needed for this test)
       4. Click **Create build configuration**
       5. Name it "Connectivity Test" and click **Create**
       6. Click **Build Steps** → **Add build step**
       7. Set **Runner type** to "Command Line"
       8. Set **Step name** to "Infrastructure Connectivity Test"
       9. Paste this script into **Custom script**:

  ```bash
  #!/bin/bash
  echo "=== Testing Infrastructure Connectivity ==="

  # Test Perforce
  echo "Testing Perforce..."
  P4PORT=$(terraform output -raw p4_server_connection_string)
  if p4 -p ${P4PORT} info; then
    echo "✓ Perforce connectivity successful"
  else
    echo "✗ Perforce connectivity failed"
    exit 1
  fi

  # Test Unity Accelerator
  echo "Testing Unity Accelerator..."
  ACCELERATOR_URL=$(terraform output -raw unity_accelerator_url | sed 's|https://||')
  if curl -f http://${ACCELERATOR_URL}:10080; then
    echo "✓ Unity Accelerator connectivity successful"
  else
    echo "✗ Unity Accelerator connectivity failed"
    exit 1
  fi

  # Test Unity License Server (expected to fail without license)
  echo "Testing Unity License Server..."
  if [ -n "${UNITY_LICENSE_SERVER_URL}" ]; then
    curl -v ${UNITY_LICENSE_SERVER_URL} || echo "⚠ License server check failed (expected without valid license)"
  else
    echo "⚠ License server not deployed"
  fi

  # Verify Unity installation
  echo "Verifying Unity installation..."
  ls -la /opt/unity/editors/*/Editor/Unity && echo "✓ Unity Editor found"

  echo "=== Infrastructure tests complete ==="

  10. Click Save
  11. Click Parameters (left sidebar) → Add new parameter
  12. Add environment variable UNITY_LICENSE_SERVER_URL with value from terraform output -raw unity_license_server_url (if deployed)
  13. Click Run in the top right corner
  14. Watch the build log - verify all connectivity checks pass (license server failure is expected without valid license)
  15. If all tests pass, the infrastructure is working correctly
  16. Clean up: Run terraform destroy to remove all resources

  Checklist

  If your change doesn't seem to apply, please leave them unchecked.

  - I have performed a self-review of this change
  - Changes have been tested
  - Changes are documented

  Breaking changes

  No major breaking changes. This is a new sample. Minor additive changes to existing modules:
  - Perforce module: Added shared_application_load_balancer_dns_name and shared_application_load_balancer_zone_id outputs for Route53 integration with P4 Swarm
  - P4 Code Review module: Added checkov skip for ElastiCache (automatic backups not required for ephemeral cache data)

  Acknowledgment

  By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

  Disclaimer: We value your time and bandwidth. As such, any pull requests created might not be successful.

gabebatista committed:
feat(samples): add Unity build pipeline with TeamCity integration

gabebatista commented on modules/teamcity/local.tf:
```diff
@@ -17,6 +17,9 @@ locals {
   efs_access_point_id = var.efs_access_point_id != null ? var.efs_access_point_id : aws_efs_access_point.teamcity_efs_data_access_point[0].id
 
+  # Service Connect namespace
+  service_connect_namespace_arn = aws_service_discovery_http_namespace.teamcity.arn
```
Good catch! Fixed.

joshral7459 commented on samples/unity-build-pipeline/docker/teamcity-unity-build-agent/Dockerfile:
```diff
@@ -0,0 +1,126 @@
+# Install Unity Editor via Unity Hub
+# Default to Unity 6 LTS - users can override with --build-arg
+ARG UNITY_VERSION=6000.0.23f1
+ARG UNITY_CHANGESET=bd20d88e54b8
```
updated editor args based on latest version https://unity.com/releases/editor/whats-new/6000.2.12f1#installs

joshral7459 commented on samples/unity-build-pipeline/docker/teamcity-unity-build-agent/Dockerfile:
```diff
@@ -0,0 +1,126 @@
+ENV UNITY_PATH=/opt/unity
+
+# Install system dependencies required by Unity and TeamCity
+RUN apt-get update && apt-get install -y \
```
I had to Add libgbm1 due to error - error while loading shared libraries: libgbm.so.1: cannot open shared object file: No such file or directory

joshral7459 commented on samples/unity-build-pipeline/README.md:
```diff
@@ -0,0 +1,1063 @@
+# Step 3: Build the image (takes 15-30 minutes)
+docker build \
+  --build-arg UNITY_VERSION=6000.0.23f1 \
+  --build-arg UNITY_CHANGESET=bd20d88e54b8 \
```
I had to modify the args based to use the newest version of the editor https://unity.com/releases/editor/whats-new/6000.2.12f1#installs

0.23f1 is labeled as a security risk

joshral7459 commented on samples/unity-build-pipeline/README.md:
```diff
@@ -0,0 +1,1063 @@
+p4 sync
+
+# Navigate to your workspace root
+cd ~/perforce/<workspace-name>/
```
My workspace root defaulted to my current directory which happened to be the unity sample folder in the toolkit. Did i miss a step to set it to something else?

joshral7459 commented on samples/unity-build-pipeline/README.md:
```diff
@@ -0,0 +1,1063 @@
+
+**Step 2.1: Initial Access**
+
+Visit `https://swarm.yourdomain.com` in your browser. On first access, Swarm will:
```
I was not able to access the swarm server at all.

joshral7459 reviewed (changes requested):
Minor issues that prevent progress. Would be happy to do a working session and step through the deployment together.

joshral7459 commented on samples/unity-build-pipeline/main.tf:
```diff
@@ -0,0 +1,162 @@
+  project_prefix = local.project_prefix
+  vpc_id         = aws_vpc.unity_pipeline_vpc.id
+
+  create_route53_private_hosted_zone      = true
```
is the intention to create a 2nd hosted zone for perforce and its subdomains as defined in the perforce module, route53 file?

gabebatista committed:
fix(samples/unity-build-pipeline): resolve markdown lint errors

- Add language specifiers to fenced code blocks
- Convert bold emphasis to proper headings for sub-steps
- Add blank lines around lists and fenced code blocks
- Fix heading spacing issues

gabebatista committed:
fix(samples/unity-build-pipeline): resolve docker README markdown lint errors

gabebatista committed:
docs(samples/unity-build-pipeline): add architecture diagram and fix deployment descriptions

- Replace ASCII diagram with draw.io architecture diagram
- Fix P4 Swarm deployment type: ECS -> EC2 with Auto Scaling Group
- Fix Unity License Server deployment type: ECS -> EC2 with ENI
- Fix troubleshooting section references
- Remove cost estimate section (too risky to maintain)

gabebatista committed:
feat(samples/unity-build-pipeline): add ALB security group rules for TeamCity and Unity Accelerator

- Add alb_security_group_id output to TeamCity module
- Add ingress rules for TeamCity ALB (port 443) in sample security.tf
- Add ingress rules for Unity Accelerator ALB (port 443) in sample security.tf
- Document security group access pattern in README

gabebatista committed:
fix(perforce): add NLB DNS outputs and fix Swarm public DNS record

- Add shared_network_load_balancer_dns_name output to Perforce module
- Add shared_network_load_balancer_zone_id output to Perforce module
- Fix P4 Swarm DNS record to point to NLB instead of internal ALB

The Swarm DNS was pointing to the internal ALB which is not accessible
from the internet. It should point to the public NLB which forwards
HTTPS traffic to the internal ALB.

gabebatista committed:
fix(perforce): add NLB security group output and Swarm access rule

- Add shared_network_load_balancer_security_group_id output
- Add security group rule for HTTPS traffic to NLB for Swarm access

gabebatista committed:
fix(ci): mitigate script injection in docs workflows

Use intermediate env variables for PR number interpolation instead of
direct ${{ }} expansion in shell scripts to prevent potential injection.

Affected files:
- docs-cleanup.yml: mike delete command
- docs-preview.yml: content change detection
